### PR TITLE
Prevent using already produced serial number

### PR DESCRIPTION
**Pull request by Whenrow**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mrp/models/mrp_abstract_workorder.py [hunk 1]
```diff
@@ -316,8 +333,21 @@ def _onchange_lot_id(self):
...
-        if self.product_id.tracking == 'serial':
+        if self.product_id.tracking == 'serial' and self.lot_id:
             self.qty_done = 1
+            # Alert the user if this serial number as already been used
+            sml = self.env['stock.move.line'].search_count([
+                ('lot_id', '=', self.lot_id.id),
+                ('location_id', '=', self._get_production().location_src_id.id),
+                ('location_dest_id', '=', self._get_production().production_location_id.id),
+                ('qty_done', '=', 1),
+                ('state', '=', 'done')
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_abstract_workorder.py]: one state done, qty = 1 the other no state and qty_done > 0

File: addons/mrp/models/mrp_workorder.py [hunk 2]
```diff
@@ -154,6 +154,9 @@ def _search_date_planned_finished(self, operator, value):
     def _onchange_finished_lot_id(self):
         """When the user changes the lot being currently produced, suggest
         a quantity to produce consistent with the previous workorders. """
+        res = super(MrpWorkorder, self)._onchange_finished_lot_id()
+        if res:
+            return res
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_workorder.py]: why?

File: addons/mrp/models/mrp_abstract_workorder.py [hunk 3]
```diff
@@ -44,6 +44,23 @@ def _workorder_line_ids(self):
...
         return self.raw_workorder_line_ids | self.finished_workorder_line_ids
 
+    @api.onchange('finished_lot_id')
+    def _onchange_finished_lot_id(self):
+        # sanity check
+        if self.product_tracking == 'serial' and self.finished_lot_id:
+            # Alert the user if this serial number as already been produced
+            sml = self.env['stock.move.line'].search_count([
+                ('lot_id', '=', self.finished_lot_id.id),
+                ('location_id', '=', self.production_id.production_location_id.id),
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_abstract_workorder.py]: what if 1 company multipe warehouses multiple production_location?

File: addons/mrp/models/mrp_workorder.py (hunk 2 above)
Comment (Whenrow) [reply addons/mrp/models/mrp_workorder.py]: because _onchange_finished_lot_id() on abstract_workorder could return a warning

Comment (sle-odoo) [reply addons/mrp/models/mrp_workorder.py]: don't shortcircuit here an make these ocnhange always return something

File: addons/mrp/models/mrp_abstract_workorder.py [hunk 4]
```diff
@@ -44,6 +44,25 @@ def _workorder_line_ids(self):
...
+        if self.product_tracking == 'serial' and self.finished_lot_id:
+            # Alert the user if this serial number as already been produced
+            sml = self.env['stock.move.line'].search_count([
+                ('lot_id', '=', self.finished_lot_id.id),
+                ('location_id.usage', '=', 'production'),
+                ('qty_done', '=', 1),
+                ('state', '=', 'done')
+            ])
+            if sml:
+                return {'warning': {
```
Comment (Whenrow) [inline addons/mrp/models/mrp_abstract_workorder.py]: test comment

File: addons/mrp/models/mrp_abstract_workorder.py [hunk 5]
```diff
@@ -362,7 +374,10 @@ def _update_move_lines(self):
             raise UserError(_('Please enter a lot or serial number for %s !' % self.product_id.display_name))
 
         if self.lot_id and self.product_id.tracking == 'serial' and self.lot_id in self.move_id.move_line_ids.filtered(lambda ml: ml.qty_done).mapped('lot_id'):
-            raise UserError(_('You cannot consume the same serial number twice. Please correct the serial numbers encoded.'))
+            if self[self._get_raw_workorder_inverse_name()]:
+                raise UserError(_('The serial number used for component %s has already been consumed' % self.product_id.name))
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_abstract_workorder.py]: wrong call to _(

File: addons/mrp/models/mrp_workorder.py [hunk 6]
```diff
@@ -384,10 +384,13 @@ def record_production(self):
             return True
 
         self.ensure_one()
+        self._check_sn_uniquiness()
         self._check_company()
         if float_compare(self.qty_producing, 0, precision_rounding=self.product_uom_id.rounding) <= 0:
             raise UserError(_('Please set the quantity you are currently producing. It should be different from zero.'))
-
+        if 'check_ids' not in self:
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_workorder.py]: we need a better way to know if mrp_workorder is installed or not :)

File: addons/mrp/models/mrp_abstract_workorder.py [hunk 7]
```diff
@@ -312,7 +324,7 @@ class MrpAbstractWorkorderLine(models.AbstractModel):
     lot_id = fields.Many2one(
         'stock.production.lot', 'Lot/Serial Number',
         check_company=True,
-        domain="[('product_id', '=', product_id), '|', ('company_id', '=', False), ('company_id', '=', company_id)]")
+        domain="[('product_id', '=', product_id), '|', ('company_id', '=', False), ('company_id', '=', parent.company_id)]")
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_abstract_workorder.py]: shoudln't this one go in v13?

File: addons/mrp/wizard/mrp_product_produce_views.xml [hunk 8]
```diff
@@ -31,7 +31,7 @@
                         <field name="raw_workorder_line_ids" attrs="{'invisible': [('raw_workorder_line_ids', '=', [])]}" nolabel="1" context="{'w_production': True, 'active_id': production_id, 'default_finished_lot_id': finished_lot_id}">
                             <tree editable="top" delete="0" decoration-danger="(qty_to_consume &lt; qty_done)">
                                 <field name="company_id" invisible="1"/>
-                                <field name="product_id" attrs="{'readonly': [('move_id', '!=', False)]}" required="1" domain="[('id', '!=', parent.product_id), ('type', 'in', ['product', 'consu']), '|', ('company_id', '=', False), ('company_id', '=', company_id)]" force_save="1"/>
+                                <field name="product_id" attrs="{'readonly': [('move_id', '!=', False)]}" required="1" domain="[('id', '!=', parent.product_id), ('type', 'in', ['product', 'consu']), '|', ('company_id', '=', False), ('company_id', '=', parent.company_id)]" force_save="1"/>
```
Comment (sle-odoo) [inline addons/mrp/wizard/mrp_product_produce_views.xml]: v13?

File: addons/mrp/wizard/mrp_product_produce_views.xml [hunk 9]
```diff
@@ -47,7 +47,7 @@
                         <field name="finished_workorder_line_ids" attrs="{'invisible': [('finished_workorder_line_ids', '=', [])]}" nolabel="1" context="{'w_production': True, 'active_id': production_id, 'default_finished_lot_id': finished_lot_id}">
                             <tree editable="top" delete="0" decoration-danger="(qty_to_consume &lt; qty_done)">
                                 <field name="company_id" invisible="1"/>
-                                <field name="product_id" attrs="{'readonly': [('move_id', '!=', False)]}" required="1" domain="[('id', '!=', parent.product_id), ('type', 'in', ['product', 'consu']), '|', ('company_id', '=', False), ('company_id', '=', company_id)]" force_save="1"/>
+                                <field name="product_id" attrs="{'readonly': [('move_id', '!=', False)]}" required="1" domain="[('id', '!=', parent.product_id), ('type', 'in', ['product', 'consu']), '|', ('company_id', '=', False), ('company_id', '=', parent.company_id)]" force_save="1"/>
```
Comment (sle-odoo) [inline addons/mrp/wizard/mrp_product_produce_views.xml]: v13?

**Comment by amoyaux** [comment]
robodoo r+

**Comment by amoyaux** [comment]
robodoo rebase-ff

**Pull request closed without merging**